data_structures/linked_list/circular_doubly_linked_list.c

```diff
@@ -0,0 +1,225 @@
+        temp = (*start)->prev;
+        n1->next = *start;
+        (*start)->prev = n1;
+        n1->prev = temp;
+        temp->next = n1;
+        *start = n1;
+    }
+    printf("\nNode succesfully inserted\n\n");
+}
+
+/**
+ * @brief   Insert the node at end of list
+ * @param   start   starting pointer of list
+ * @param   data    the data which we want to insert into list
+ * @returns void
+ */
+void insert_at_last(struct node **start, uint64_t data)
+{
+    struct node *n1;
+    n1 = (struct node *)malloc(sizeof(struct node));
+    n1->value = data;
+    if (*start == NULL)
+    {
+        n1->next = n1;
+        n1->prev = n1;
+        *start = n1;
+    }
+    else
+    {
+        struct node *temp1, *temp2;
+        temp1 = *start;
+        temp2 = (*start)->prev;
+        n1->prev = temp2;
+        n1->next = temp1;
+        temp1->prev = n1;
+        temp2->next = n1;
+    }
+    printf("\nNode succesfully inserted\n\n");
+}
+
+/**
+ * @brief   Deletion of the first node in list function
+ * @param   start   starting pointer of list
+ * @returns void
+ */
+void delete_first_node(struct node **start)
+{
+    if (*start == NULL)
+    {
+        printf("⚠️list is Empty⚠️\n");
+    }
+    else
+    {
+        struct node *temp1, *temp2;
+        temp1 = *start;
+        temp2 = temp1->prev;
+        if (temp1 == temp2)
+        {
+            free(temp2);
+            *start = NULL;
+        }
+        else
+        {
+            temp2->next = temp1->next;
+            (temp1->next)->prev = temp2;
+            *start = temp1->next;
+            free(temp1);
+            temp1 = NULL;
+        }
+        printf("\nNode succesfully deleted\n\n");
+    }
+}
+
+/**
+ * @brief   Deletion of the last node in list function
+ * @param   start   starting pointer of list
+ * @returns void
+ */
+void delete_last_node(struct node **start)
+{
+    if (*start == NULL)
+    {
+        printf("list is Empty\n");
+    }
+    else
+    {
+        struct node *temp1, *temp2;
+        temp1 = *start;
+        temp2 = temp1->prev;
+        if (temp1 == temp2)
+        {
+            free(temp2);
+            *start = NULL;
+        }
+        else
+        {
+            (temp2->prev)->next = temp1;
+            temp1->prev = temp2->prev;
+            free(temp2);
+            temp2 = NULL;
+        }
+        printf("\nNode succesfully deleted\n\n");
+    }
+}
+
+/**
+ * @brief   Display list function
+ * @param   start   starting pointer of list
+ * @returns void
+ */
+void Display(struct node *start)
+{
+    printf("\nIts Your Linked-list\n\n");
+    struct node *temp;
+    temp = start;
+    if (start != NULL)
+    {
+        while (temp->next != start)
+        {
+            printf("%lld -> ", temp->value);
+            temp = temp->next;
+        }
+        if (temp->next == start)
+        {
+            printf("%lld", temp->value);
+        }
+    }
+    else
+    {
+        printf("list is Empty\n");
+    }
+    printf("\n");
+}
+
```

Comment:
I think the self-test implementations are still missing. Do you need any help with this?

Reply:
Yes...How to do self-test implementations...You told me to use `assert ` Can you show me how to use assert in my particular program at least for one function?

Reply:
This is a code example using `assert`.

```c++
#include <iostream>
#include <cassert>

int sum_function(int var)
{
    return var += 2; // or `var = var + 2`
}

int main()
{
    int my_var = 7;
    // If `my_var` is 9 after the change, the code will continue.
    // Otherwise, the program will stop.
    assert(sum_function(my_var) == 9);
}
```

Reply:
I have tried on many compilers but they are giving an error like,
```
Line 4: Char 10: fatal error: cassert: No such file or directory [solution.c]
 #include <cassert>
          ^~~~~~~~~
compilation terminated.
```

But I think the code is working just fine. You can check out it's working [here](https://leetcode.com/playground/WcRBMWa8).

Reply:
Whoops, got confused, should be `assert.h`. 😅
Would be great if you could make the changes here.

Reply:
I think I cannot assert the void function because when I try to assert the delete node function it's giving error. 
```
Finished in N/A
In file included from prog_joined.c:7:
solution.c: In function ‘main’
Line 219: Char 12: error: void value not ignored as it ought to be [solution.c]
     assert(delete_first_node(&start))
```
But I have asserted the reverse linked list function. In the same file, I shared earlier.

Reply:
You might need to convert it to a returnable value such as `int` or the linked list.

> You can check out it's working [here](https://leetcode.com/playground/WcRBMWa8).

Is that the updated code or the current code in this PR? 🤔

Reply:
Yeah, It's updated, I have removed the menu driven thing and added the reverse linked list function to it.